internal/controllers/insights.go

```diff
@@ -0,0 +1,376 @@
+	}
+
+	return r.createOrUpdateProxyService(ctx, svc, owner)
+}
+
+func (r *InsightsReconciler) getTokenFromPullSecret(ctx context.Context) (*string, error) {
+	// Get the global pull secret
+	pullSecret := &corev1.Secret{}
+	err := r.Client.Get(ctx, types.NamespacedName{Namespace: "openshift-config", Name: "pull-secret"}, pullSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	// Look for the .dockerconfigjson key within it
+	dockerConfigRaw, pres := pullSecret.Data[corev1.DockerConfigJsonKey]
+	if !pres {
+		return nil, fmt.Errorf("no %s key present in pull secret", corev1.DockerConfigJsonKey)
+	}
+
+	// Unmarshal the .dockerconfigjson into a struct
+	dockerConfig := struct {
+		Auths map[string]struct {
+			Auth string `json:"auth"`
+		} `json:"auths"`
+	}{}
+	err = json.Unmarshal(dockerConfigRaw, &dockerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Look for the "cloud.openshift.com" auth
+	openshiftAuth, pres := dockerConfig.Auths["cloud.openshift.com"]
+	if !pres {
+		return nil, errors.New("no \"cloud.openshift.com\" auth within pull secret")
+	}
+
+	token := strings.TrimSpace(openshiftAuth.Auth)
+	if strings.Contains(token, "\n") || strings.Contains(token, "\r") {
+		return nil, fmt.Errorf("invalid cloud.openshift.com token")
+	}
+	return &token, nil
+}
+
+func (r *InsightsReconciler) getUserAgentString(ctx context.Context) (*string, error) {
+	cv := &configv1.ClusterVersion{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: "version"}, cv)
+	if err != nil {
+		return nil, err
+	}
+
+	userAgent := fmt.Sprintf("%s cluster/%s", r.UserAgentPrefix, cv.Spec.ClusterID)
+	return &userAgent, nil
+}
+
+func (r *InsightsReconciler) createOrUpdateProxySecret(ctx context.Context, secret *corev1.Secret, owner metav1.Object,
+	config string) error {
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, secret, func() error {
+		// Set the config map as controller
+		if err := controllerutil.SetControllerReference(owner, secret, r.Scheme); err != nil {
+			return err
+		}
+		// Add the APICast config.json
+		if secret.StringData == nil {
+			secret.StringData = map[string]string{}
+		}
+		secret.StringData["config.json"] = config
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	r.Log.Info(fmt.Sprintf("Secret %s", op), "name", secret.Name, "namespace", secret.Namespace)
+	return nil
+}
+
+func (r *InsightsReconciler) createOrUpdateProxyDeployment(ctx context.Context, deploy *appsv1.Deployment, owner metav1.Object) error {
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, deploy, func() error {
+		labels := map[string]string{"app": common.ProxyDeploymentName}
+		annotations := map[string]string{}
+		common.MergeLabelsAndAnnotations(&deploy.ObjectMeta, labels, annotations)
+		// Set the config map as controller
+		if err := controllerutil.SetControllerReference(owner, deploy, r.Scheme); err != nil {
+			return err
+		}
+		// Immutable, only updated when the deployment is created
+		if deploy.CreationTimestamp.IsZero() {
```

Comment:
Yes, I guess that usually should not happen when the resource is managed by us. So if we create a resource and compare the desired with actual there is no reason why should be this different (except if it is our change in the code). But yes, since cannot change, might event not make sense to compare, or if yes just log a warning that it is different.